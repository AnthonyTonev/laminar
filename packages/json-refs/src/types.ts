export interface JsonObject {
  $ref?: string;
  $id?: string;
  id?: string;
  [key: string]: unknown;
}

export type JsonPointerObject = JsonObject | boolean;

export interface RefSchema {
  $ref: string;
  [key: string]: unknown;
}
export interface RefMap {
  [ref: string]: JsonPointerObject;
}

export interface TraversableJsonObject {
  [key: string]: JsonPointerObject;
}

export interface Context {
  schema: JsonPointerObject;
  refs: RefMap;
}

export interface FileContext {
  parentId?: string;
  cwd?: string;
  uris?: string[];
}

export interface LoadedJsonObject {
  uri: string;
  content: JsonPointerObject;
  cwd?: string;
}

export interface ResolvedJsonObject {
  schema: JsonPointerObject;
  refs: RefMap;
  uris: string[];
}
