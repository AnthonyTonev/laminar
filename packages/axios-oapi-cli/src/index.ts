export { convertSchema } from './convert-schema';
export { oapiTs, schemaTs } from './convert';
