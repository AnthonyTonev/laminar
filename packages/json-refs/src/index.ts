import fetch from 'node-fetch';
import { URL } from 'url';

interface IdSchema {
  $id: string;
}
interface RefSchema {
  $ref: string;
}
interface TraversableSchema {
  [key: string]: any;
}

interface FilesMap {
  [file: string]: any;
}

interface Context {
  schema: any;
  files: FilesMap;
}

export const isTraversable = (schema: any): schema is TraversableSchema =>
  schema && typeof schema === 'object';

export const isIdSchema = (schema: any): schema is IdSchema =>
  isTraversable(schema) && '$id' in schema && schema.$id;

export const isRefSchema = (schema: any): schema is RefSchema =>
  isTraversable(schema) && '$ref' in schema && schema.$ref;

export const currentId = (schema: any, parentId?: string) =>
  isIdSchema(schema) ? new URL(schema.$id, parentId).toString() : parentId;

export const currentUrl = (url: string | undefined, id?: string) =>
  url ? new URL(url, id).toString() : undefined;

export const reduceSchema = <TResult = any>(
  schema: any,
  cb: (all: TResult, item: any, id?: string) => TResult,
  initial: TResult,
  id?: string,
): TResult =>
  isTraversable(schema)
    ? Object.values(schema).reduce(
        (all, item) => reduceSchema(item, cb, all, currentId(schema, id)),
        cb(initial, schema, id),
      )
    : cb(initial, schema, id);

export const extractNamedRefs = (document: any): FilesMap =>
  reduceSchema(
    document,
    (all, item, id) =>
      isIdSchema(item) ? { ...all, [new URL(item.$id, id).toString()]: item } : all,
    {},
  );

export const extractUrls = (schema: any, namedRefs: string[] = []) =>
  reduceSchema<string[]>(
    schema,
    (all, item, id) => {
      if (isRefSchema(item)) {
        const [url] = item.$ref.split('#');
        const fullUrl = currentUrl(url, id);

        if (fullUrl && !all.includes(fullUrl) && !namedRefs.includes(fullUrl)) {
          return [...all, fullUrl];
        }
      }
      return all;
    },
    [],
  );

export const extractFiles = async (schema: any): Promise<FilesMap> => {
  const refs = extractNamedRefs(schema);
  const result = await Promise.all(
    extractUrls(schema, Object.keys(refs)).map(url =>
      fetch(url)
        .then(response => response.json())
        .then(content => extractFiles(content).then(filesMap => ({ ...filesMap, [url]: content }))),
    ),
  );
  return result.reduce((all, item) => ({ ...all, ...item }), refs);
};

export const parseJsonPointer = (name: string) =>
  decodeURIComponent(name.replace('~1', '/').replace('~0', '~'));

export const getJsonPointer = (document: any, pointer: string) =>
  pointer
    .split('/')
    .reduce<any>(
      (item, name) => (name ? (item ? item[parseJsonPointer(name)] : undefined) : item),
      document,
    );

export const resolveNestedRefs = (schema: any, context: Context, parentId?: string) => {
  const id = currentId(schema, parentId);

  if (isTraversable(schema)) {
    for (const [key, item] of Object.entries(schema)) {
      schema[key] = resolveNestedRefs(item, context, id);
    }
  }

  if (isRefSchema(schema)) {
    const [url, pointer] = schema.$ref.split('#');
    const fullUrl = currentUrl(url, id);
    const currentDocument = fullUrl ? context.files[fullUrl] : context.schema;

    return pointer ? getJsonPointer(currentDocument, pointer) : currentDocument;
  }

  return schema;
};

export const resolveRefs = async (schema: any): Promise<any> =>
  resolveNestedRefs(schema, { schema, files: await extractFiles(schema) });
