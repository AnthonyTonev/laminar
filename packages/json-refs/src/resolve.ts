import { existsSync, readFileSync } from 'fs';
import * as YAML from 'js-yaml';
import fetch from 'node-fetch';
import { dirname, join } from 'path';
import { URL } from 'url';
import {
  JsonPointerObject,
  TraversableJsonObject,
  RefSchema,
  FileContext,
  RefMap,
  LoadedJsonObject,
  Context,
  ResolvedJsonObject,
} from './types';
import { ResolveError } from './ResolveError';

export const isTraversable = (schema: unknown): schema is TraversableJsonObject =>
  schema && typeof schema === 'object';

export const toUrl = (url: string, base?: string): string | undefined => {
  try {
    return new URL(url, base).toString();
  } catch (error) {
    return undefined;
  }
};

export const getId = (schema: JsonPointerObject): string | undefined => {
  if (isTraversable(schema)) {
    if ('$id' in schema && schema.$id && typeof schema.$id === 'string') {
      return schema.$id;
    } else if ('id' in schema && schema.id && typeof schema.id === 'string') {
      return schema.id;
    }
  }
  return undefined;
};

export const isSchema = (schema: unknown): schema is JsonPointerObject =>
  typeof schema === 'boolean' || (typeof schema === 'object' && schema !== null);

export const isRefSchema = (schema: JsonPointerObject): schema is RefSchema =>
  isTraversable(schema) &&
  '$ref' in schema &&
  schema.$ref !== undefined &&
  typeof schema.$ref === 'string';

export const currentId = (schema: JsonPointerObject, parentId?: string): string | undefined => {
  const id = getId(schema);
  return id ? toUrl(id, parentId) : parentId;
};

export const currentUrl = (
  url?: string,
  { cwd, parentId }: FileContext = {},
): string | undefined => {
  const fullUrl = url && toUrl(url, parentId);
  if (fullUrl) {
    return fullUrl;
  } else if (url && cwd) {
    if (!existsSync(join(cwd, url))) {
      throw new ResolveError(`File ${url} in ${cwd} does not exist`);
    }
    return url;
  } else {
    return undefined;
  }
};

export const reduceSchema = <TResult = JsonPointerObject>(
  schema: JsonPointerObject,
  cb: (all: TResult, item: JsonPointerObject, id?: string) => TResult,
  initial: TResult,
  id?: string,
): TResult =>
  isTraversable(schema)
    ? Object.values(schema).reduce(
        (all, item) => reduceSchema(item, cb, all, currentId(schema, id)),
        cb(initial, schema, id),
      )
    : cb(initial, schema, id);

export const extractNamedRefs = (document: JsonPointerObject): RefMap =>
  reduceSchema(
    document,
    (all, item, id) => {
      const itemId = getId(item);
      const url = itemId ? toUrl(itemId, id) : false;
      return url ? { ...all, [url]: item } : all;
    },
    {},
  );

export const extractUrls = (
  schema: JsonPointerObject,
  namedRefs: string[] = [],
  fileContext: FileContext = {},
): string[] =>
  reduceSchema<string[]>(
    schema,
    (all, item, parentId) => {
      if (isRefSchema(item)) {
        const [url] = item.$ref.split('#');
        const fullUrl = currentUrl(url, { ...fileContext, parentId });

        if (fullUrl && !all.includes(fullUrl) && !namedRefs.includes(fullUrl)) {
          return [...all, fullUrl];
        }
      }
      return all;
    },
    [],
  );

export const loadFile = async (
  uri: string,
  { cwd }: FileContext = {},
): Promise<LoadedJsonObject> => {
  const url = toUrl(uri);
  if (url) {
    const result = await fetch(uri);
    if (result.headers.get('content-type') === 'application/yaml') {
      return { uri, content: YAML.load(await result.text()) };
    } else {
      return { uri, content: await result.json() };
    }
  } else {
    const file = cwd ? join(cwd, uri) : uri;
    const content = readFileSync(file, 'utf8');
    const newCwd = dirname(file);

    if (uri.endsWith('.yaml') || uri.endsWith('.yml')) {
      return { uri: `file://${file}`, content: YAML.load(content), cwd: newCwd };
    } else {
      return { uri: `file://${file}`, content: JSON.parse(content), cwd: newCwd };
    }
  }
};

export const parseJsonPointer = (name: string): string =>
  decodeURIComponent(name.replace('~1', '/').replace('~0', '~'));

export const getJsonPointer = (
  document: JsonPointerObject,
  pointer: string,
): JsonPointerObject | undefined =>
  pointer
    .split('/')
    .reduce<JsonPointerObject | undefined>(
      (item, name) =>
        name ? (isTraversable(item) ? item[parseJsonPointer(name)] : undefined) : item,
      document,
    );

export const resolveNestedRefs = (
  schema: JsonPointerObject,
  context: Context,
  fileContext: FileContext = {},
): JsonPointerObject => {
  const parentId = currentId(schema, fileContext.parentId);

  if (isTraversable(schema)) {
    for (const [key, item] of Object.entries(schema)) {
      schema[key] = resolveNestedRefs(item, context, { ...fileContext, parentId });
    }
  }

  if (isRefSchema(schema)) {
    const [url, pointer] = schema.$ref.split('#');

    const fullUrl = currentUrl(url, { ...fileContext, parentId });
    const fullPointer = [fullUrl, pointer].join('#');
    if (!context.refs[fullPointer]) {
      const currentDocument = fullUrl ? context.refs[fullUrl] : context.schema;
      const newContent = pointer ? getJsonPointer(currentDocument, pointer) : currentDocument;
      if (newContent !== undefined) {
        context.refs[fullPointer] = newContent;
      }
    }

    return { $ref: fullPointer };
  }

  return schema;
};

export const extractFiles = async (
  schema: JsonPointerObject,
  options: FileContext = {},
): Promise<{ refs: RefMap; uris: string[] }> => {
  const initialRefs = extractNamedRefs(schema);
  const result = await Promise.all(
    extractUrls(schema, Object.keys(initialRefs), options).map(async (url) => {
      const { content, cwd, uri } = await loadFile(url, options);
      const { refs, uris } = await extractFiles(content, { cwd });
      const ref = resolveNestedRefs(content, { schema: content, refs }, { cwd });
      return { uris: [...uris, uri], refs: { [url]: ref, ...refs } };
    }),
  );

  return result.reduce(
    (all, item) => ({ uris: [...all.uris, ...item.uris], refs: { ...all.refs, ...item.refs } }),
    { uris: [], refs: initialRefs },
  );
};

export const resolveRefs = async (
  original: JsonPointerObject,
  fileContext: FileContext = {},
): Promise<ResolvedJsonObject> => {
  const copy: JsonPointerObject = JSON.parse(JSON.stringify(original));
  const { refs, uris } = await extractFiles(copy, fileContext);
  const context = { schema: copy, refs, uris };
  const schema = resolveNestedRefs(copy, context, fileContext);
  return { schema, refs, uris };
};

export const resolveRefsFile = async (file: string): Promise<ResolvedJsonObject> => {
  const { content, cwd, uri } = await loadFile(file);
  const resolved = await resolveRefs(content, { cwd });
  return { ...resolved, uris: [uri, ...resolved.uris] };
};
