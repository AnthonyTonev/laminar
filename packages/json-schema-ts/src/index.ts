const parse = (schema: any, refs: { [key: string]: string }) => {
  if (schema.type === 'string' || 'pattern' in schema) {
    return 'string';
  }

  if (schema.type === 'integer' || schema.type === 'number') {
    return 'number';
  }
};
