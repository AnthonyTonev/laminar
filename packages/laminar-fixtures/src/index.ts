export {
  generateIdSequnce,
  cloneFixture,
  fixture,
  multiFixture,
  template,
  id,
  rel,
  alternate,
  generate,
} from './fixtures';
export { toTeardownQueries, toSetupQueries } from './queries';
export { tearDown, setUp } from './run';
