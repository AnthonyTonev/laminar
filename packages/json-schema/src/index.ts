import { resolveRefs } from '@ovotech/json-refs';
import { validateSchema } from './helpers';
import { messages } from './messages';
import { Schema, ValidateOptions } from './types';
import { validateAdditionalProperties } from './validators/validateAdditionalProperties';
import { validateAllOf } from './validators/validateAllOf';
import { validateAnyOf } from './validators/validateAnyOf';
import { validateBoolean } from './validators/validateBoolean';
import { validateConditionals } from './validators/validateConditionals';
import { validateConst } from './validators/validateConst';
import { validateContains } from './validators/validateContains';
import { validateDependencies } from './validators/validateDependencies';
import { validateEnum } from './validators/validateEnum';
import { validateExclusiveMaximum } from './validators/validateExclusiveMaximum';
import { validateExclusiveMinimum } from './validators/validateExclusiveMinimum';
import { validateFormat } from './validators/validateFormat';
import { validateItems } from './validators/validateItems';
import { validateMaximum } from './validators/validateMaximum';
import { validateMaxItems } from './validators/validateMaxItems';
import { validateMaxLength } from './validators/validateMaxLength';
import { validateMaxProperties } from './validators/validateMaxProperties';
import { validateMinimum } from './validators/validateMinimum';
import { validateMinItems } from './validators/validateMinItems';
import { validateMinLength } from './validators/validateMinLength';
import { validateMinProperties } from './validators/validateMinProperties';
import { validateMultipleOf } from './validators/validateMultipleOf';
import { validateNot } from './validators/validateNot';
import { validateOneOf } from './validators/validateOneOf';
import { validatePattern } from './validators/validatePattern';
import { validatePatternProperties } from './validators/validatePatternProperties';
import { validateProperties } from './validators/validateProperties';
import { validatePropertyNames } from './validators/validatePropertyNames';
import { validateRequired } from './validators/validateRequired';
import { validateType } from './validators/validateType';
import { validateUniqueItems } from './validators/validateUniqueItems';

export {
  childOptions,
  isJsonSchema,
  isObject,
  isEqual,
  isUniqueWith,
  NoErrors,
  HasError,
  HasErrors,
  CombineResults,
  validateSchema,
} from './helpers';

export { Schema, JsonSchema, ValidateOptions, Validator, Messages } from './types';

export const draft7 = [
  validateBoolean,
  validateConst,
  validateEnum,
  validateType,
  validateFormat,
  validateMultipleOf,
  validateMinimum,
  validateExclusiveMinimum,
  validateMaximum,
  validateExclusiveMaximum,
  validatePattern,
  validateMinLength,
  validateMaxLength,
  validateMinProperties,
  validateMaxProperties,
  validateRequired,
  validateProperties,
  validatePatternProperties,
  validatePropertyNames,
  validateDependencies,
  validateAdditionalProperties,
  validateMinItems,
  validateMaxItems,
  validateUniqueItems,
  validateItems,
  validateContains,
  validateConditionals,
  validateNot,
  validateOneOf,
  validateAllOf,
  validateAnyOf,
];

export const validate = async (
  original: Schema,
  value: any,
  { name = 'value', validators = draft7 }: Partial<ValidateOptions> = {},
) => {
  const schema = await resolveRefs(original);
  const result = validateSchema(schema, value, { name, validators });
  const errors = result.errors.map(error => messages[error.code](error));

  return { schema, errors, valid: errors.length === 0 };
};
