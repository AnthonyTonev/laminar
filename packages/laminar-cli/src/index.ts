import { ResolvedSchema } from '@ovotech/json-schema';
import { printDocument } from '@ovotech/ts-compose';
import { OpenAPIObject, SchemaObject } from 'openapi3-ts';
import { convertOapi as apiConvert } from './commands/api/convert';
import { convertOapi as axiosConvert } from './commands/axios/convert';
import { convertSchema } from './convert-schema';

export { convertSchema } from './convert-schema';

export const apiContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(apiConvert({ root: schema as OpenAPIObject, refs }, schema as OpenAPIObject));

export const axiosContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(axiosConvert({ root: schema as OpenAPIObject, refs }, schema as OpenAPIObject));

export const schemaContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(convertSchema({ root: schema as SchemaObject, refs }, schema));
