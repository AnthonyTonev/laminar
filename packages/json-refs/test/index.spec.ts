import * as nock from 'nock';
import { extractFiles, extractNamedRefs, extractUrls, resolveRefs, JsonObject } from '../src';

describe('json-refs', () => {
  it.each<[string, JsonObject, string[]]>([
    [
      'no ref',
      {
        $ref: '#/definition/test',
        definition: {
          test: { maximum: 10 },
        },
      },
      [],
    ],
    [
      'single ref',
      {
        $ref: 'http://one.test#/test',
      },
      ['http://one.test/'],
    ],
    [
      'nested multiple ref',
      {
        test: { $ref: 'http://one.test#/test' },
        other: {
          deep: { $ref: 'http://two.test#/test' },
        },
      },
      ['http://one.test/', 'http://two.test/'],
    ],
    [
      'nested multiple ref with ids',
      {
        $id: 'http://one.test',
        test: { $ref: 'first#/test' },
        other: {
          deep: { $ref: 'second#/test' },
        },
      },
      ['http://one.test/first', 'http://one.test/second'],
    ],
    [
      'nested multiple ref with ids old style',
      {
        id: 'http://one.test',
        test: { $ref: 'first#/test' },
        other: {
          deep: { $ref: 'second#/test' },
        },
      },
      ['http://one.test/first', 'http://one.test/second'],
    ],
    [
      'nested multiple refs and different ids',
      {
        one: {
          $id: 'http://one.test',
          test: { $ref: 'first#/test-1' },
        },
        two: {
          $id: 'http://two.test',
          deep: {
            other: { $ref: 'second#/test-2' },
          },
        },
        standAlone: { $ref: 'http://three.test' },
      },
      ['http://one.test/first', 'http://two.test/second', 'http://three.test/'],
    ],
  ])('Should extractUrls for %s', (_, schema, expected) => {
    expect(extractUrls(schema)).toEqual(expected);
  });

  it.each<[string, JsonObject, JsonObject]>([
    [
      'single id',
      {
        $id: 'http://one.test',
        minimum: 12,
      },
      { 'http://one.test/': { $id: 'http://one.test', minimum: 12 } },
    ],
    [
      'nested id',
      {
        $id: 'http://one.test',
        minimum: 12,
        deep: {
          $id: 'other',
          maximum: 22,
        },
      },
      {
        'http://one.test/': {
          $id: 'http://one.test',
          minimum: 12,
          deep: { $id: 'other', maximum: 22 },
        },
        'http://one.test/other': { $id: 'other', maximum: 22 },
      },
    ],
  ])('Should extractNamedRefs for %s', (_, schema, expected) => {
    expect(extractNamedRefs(schema)).toEqual(expected);
  });

  it('Should extract multiple urls', async () => {
    const schema = {
      test: { $ref: 'http://one.test#/test' },
      other: {
        deep: { $ref: 'http://two.test/folder#/test' },
      },
      last: {
        $id: 'http://two.test',
        deep: {
          $ref: 'folder',
        },
      },
    };

    nock('http://one.test').get('/').reply(200, { $ref: 'http://three.test#' });

    nock('http://two.test').get('/folder').reply(200, { test: 2 });

    nock('http://three.test').get('/').reply(200, { $ref: 'http://four.test#/test', test: 3 });

    nock('http://four.test').get('/').reply(200, { test: 4 });

    const expected = {
      refs: {
        'http://two.test/': { $id: 'http://two.test', deep: { $ref: 'folder' } },
        'http://four.test/': { test: 4 },
        'http://four.test/#/test': 4,
        'http://three.test/': { $ref: 'http://four.test/#/test' },
        'http://three.test/#': { $ref: 'http://four.test/#/test' },
        'http://one.test/': { $ref: 'http://three.test/#' },
        'http://two.test/folder': { test: 2 },
      },
      uris: [
        'http://four.test/',
        'http://three.test/',
        'http://one.test/',
        'http://two.test/folder',
      ],
    };

    expect(await extractFiles(schema)).toEqual(expected);
  });

  it('Should resolve simple nested refs', async () => {
    const schema = {
      definitions: {
        a: { type: 'integer' },
        b: { $ref: '#/definitions/a' },
        c: { $ref: '#/definitions/b' },
      },
      $ref: '#/definitions/c',
    };

    const expected = {
      schema: {
        $ref: '#/definitions/c',
      },
      refs: {
        '#/definitions/a': { type: 'integer' },
        '#/definitions/b': { $ref: '#/definitions/a' },
        '#/definitions/c': { $ref: '#/definitions/b' },
      },
      uris: [],
    };

    expect(await resolveRefs(schema)).toEqual(expected);
  });

  it('Should resolve nested recursive with files', async () => {
    const schema: JsonObject = {
      $id: 'http://localhost:1234/tree',
      description: 'tree of nodes',
      type: 'object',
      properties: {
        meta: { type: 'string' },
        nodes: {
          type: 'array',
          items: { $ref: 'node' },
        },
      },
      required: ['meta', 'nodes'],
      definitions: {
        node: {
          $id: 'http://localhost:1234/node',
          description: 'node',
          type: 'object',
          properties: {
            value: { type: 'number' },
            subtree: { $ref: 'tree' },
          },
          required: ['value'],
        },
      },
    };

    expect(await resolveRefs(schema)).toMatchSnapshot();
  });

  it('Should resolve complex refs', async () => {
    const schema: JsonObject = {
      type: 'object',
      properties: {
        mangled: { $ref: '#/definitions/unknown/more-unknown' },
        user: {
          type: 'object',
          properties: {
            name: { $ref: '#/definitions/title' },
            id: { $ref: '#/definitions/id' },
            address: { $ref: '#/definitions/address' },
          },
        },
      },
      definitions: {
        title: {
          type: 'string',
          pattern: '^[A-Z]',
        },
        address: {
          type: 'object',
          properties: {
            id: { $ref: '#/definitions/id' },
            street: { $ref: '#/definitions/title' },
            price: { $ref: '#/definitions/test~1slash' },
            town: { $ref: '#/definitions/test~0tilde' },
            num: { $ref: '#/definitions/test%25percent' },
          },
        },
        'test/slash': {
          type: 'number',
        },
        'test~tilde': {
          type: 'string',
        },
        'test%percent': {
          type: 'integer',
        },
        id: {
          type: 'string',
          minLength: 3,
          maxLength: 4,
        },
      },
    };

    expect(await resolveRefs(schema)).toMatchSnapshot();
  });

  it('Should resolve url refs', async () => {
    nock('http://example.test')
      .get('/schema')
      .reply(200, { test: { type: 'number' } });

    const schema = { test: '123', other: { $ref: 'http://example.test/schema#/test' } };
    expect(await resolveRefs(schema)).toEqual({
      schema: { other: { $ref: 'http://example.test/schema#/test' }, test: '123' },
      refs: {
        'http://example.test/schema': {
          test: {
            type: 'number',
          },
        },
        'http://example.test/schema#/test': {
          type: 'number',
        },
      },
      uris: ['http://example.test/schema'],
    });
  });

  it('Should resolve local refs from json', async () => {
    const schema = {
      test: { $ref: 'assets/test.json#/UserResponse' },
    };

    expect(await resolveRefs(schema, { cwd: __dirname })).toEqual({
      schema: {
        test: { $ref: 'assets/test.json#/UserResponse' },
      },
      refs: {
        'assets/test.json': {
          UserResponse: {
            type: 'object',
            properties: { id: { type: 'string' }, name: { type: 'string' } },
          },
        },
        'assets/test.json#/UserResponse': {
          type: 'object',
          properties: { id: { type: 'string' }, name: { type: 'string' } },
        },
      },
      uris: [expect.stringContaining('file://')],
    });
  });

  it('Should resolve local refs from yaml', async () => {
    const schema = {
      test: { $ref: 'assets/test.yaml#/UserResponse' },
    };

    expect(await resolveRefs(schema, { cwd: __dirname })).toEqual({
      schema: {
        test: { $ref: 'assets/test.yaml#/UserResponse' },
      },
      refs: {
        'assets/test.yaml': {
          UserResponse: {
            type: 'object',
            properties: { id: { type: 'string' }, name: { type: 'string' } },
          },
        },
        'assets/test.yaml#/UserResponse': {
          type: 'object',
          properties: { id: { type: 'string' }, name: { type: 'string' } },
        },
      },
      uris: [expect.stringContaining('file://')],
    });
  });

  it('Should resolve local refs in nested files', async () => {
    const schema = {
      test: { $ref: 'assets/test.yml#/UserResponse' },
    };

    expect(await resolveRefs(schema, { cwd: __dirname })).toEqual({
      schema: {
        test: { $ref: 'assets/test.yml#/UserResponse' },
      },
      refs: {
        'test.json': {
          UserResponse: {
            type: 'object',
            properties: { id: { type: 'string' }, name: { type: 'string' } },
          },
        },
        'test.json#/UserResponse': {
          type: 'object',
          properties: { id: { type: 'string' }, name: { type: 'string' } },
        },
        'assets/test.yml': {
          UserResponse: {
            type: 'object',
            properties: { id: { $ref: 'test.json#/UserResponse' } },
          },
        },
        'assets/test.yml#/UserResponse': {
          type: 'object',
          properties: { id: { $ref: 'test.json#/UserResponse' } },
        },
      },
      uris: [expect.stringContaining('file://'), expect.stringContaining('file://')],
    });
  });

  it('Should resolve multiple refs with one load', async () => {
    nock('http://example.test')
      .get('/schema')
      .reply(200, { test: { type: 'number' }, other: { type: 'array' } });

    const schema = {
      test: { $ref: 'http://example.test/schema#/test' },
      other: { $ref: 'http://example.test/schema#/other' },
    };

    expect(await resolveRefs(schema)).toEqual({
      schema: {
        test: { $ref: 'http://example.test/schema#/test' },
        other: { $ref: 'http://example.test/schema#/other' },
      },
      refs: {
        'http://example.test/schema': { test: { type: 'number' }, other: { type: 'array' } },
        'http://example.test/schema#/test': { type: 'number' },
        'http://example.test/schema#/other': { type: 'array' },
      },
      uris: ['http://example.test/schema'],
    });
  });

  it('Should resolve nested refs', async () => {
    nock('http://example.test')
      .get('/schema')
      .reply(200, { test: { type: 'number' }, other: { type: 'array' } })
      .get('/other-schema')
      .reply(200, { last: { type: 'object' } });

    const schema = {
      test: { $ref: 'http://example.test/schema#/test' },
      other: { $ref: 'http://example.test/schema#/other' },
      last: { $ref: 'http://example.test/other-schema#/last' },
    };

    expect(await resolveRefs(schema)).toEqual({
      schema: {
        test: { $ref: 'http://example.test/schema#/test' },
        other: { $ref: 'http://example.test/schema#/other' },
        last: { $ref: 'http://example.test/other-schema#/last' },
      },
      refs: {
        'http://example.test/schema': { test: { type: 'number' }, other: { type: 'array' } },
        'http://example.test/other-schema': { last: { type: 'object' } },
        'http://example.test/schema#/test': { type: 'number' },
        'http://example.test/schema#/other': { type: 'array' },
        'http://example.test/other-schema#/last': { type: 'object' },
      },
      uris: ['http://example.test/schema', 'http://example.test/other-schema'],
    });
  });

  it('Should resolve recursive copy', async () => {
    const schema = {
      properties: {
        foo: { $ref: '#' },
      },
      additionalProperties: false,
    };

    const schema1 = await resolveRefs(schema);
    const schema2 = await resolveRefs(schema);

    expect(schema1).toEqual(schema2);
  });
});
