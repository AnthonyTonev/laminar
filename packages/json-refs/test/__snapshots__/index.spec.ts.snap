// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json-refs Should resolve nested refs 1`] = `
Object {
  "$id": "http://localhost:1234/tree",
  "definitions": Object {
    "node": Object {
      "$id": "http://localhost:1234/node",
      "description": "node",
      "properties": Object {
        "subtree": [Circular],
        "value": Object {
          "type": "number",
        },
      },
      "required": Array [
        "value",
      ],
      "type": "object",
    },
  },
  "description": "tree of nodes",
  "properties": Object {
    "meta": Object {
      "type": "string",
    },
    "nodes": Object {
      "items": Object {
        "$id": "http://localhost:1234/node",
        "description": "node",
        "properties": Object {
          "subtree": [Circular],
          "value": Object {
            "type": "number",
          },
        },
        "required": Array [
          "value",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": Array [
    "meta",
    "nodes",
  ],
  "type": "object",
}
`;
